### PR TITLE
Issu #983: Add v2 of import_catalog API endpoint with DataVersion

### PR DESCRIPTION
**Pull request by Vinai**
The v1 import_catalog API endpoint is still supported.  
This PR also includes a lot of cleanup based on inspections from PHPStorm.

### CODE REVIEW

File: src/Import/ImportCatalogCommand.php [hunk 1]
```diff
@@ -42,14 +42,14 @@ public function toMessage(): Message
 
     /**
      * @param Message $message
-     * @return static
+     * @return self
      */
```
Comment (tim-bezhashvyly) [inline src/Import/ImportCatalogCommand.php]: This DocBlock can now be removed.

File: src/Import/RestApi/CatalogImportApiV1PutRequestHandler.php [hunk 2]
```diff
@@ -5,90 +5,12 @@
...
-use LizardsAndPumpkins\Messaging\Command\CommandQueue;
-use LizardsAndPumpkins\RestApi\ApiRequestHandler;
 use LizardsAndPumpkins\Http\HttpRequest;
-use LizardsAndPumpkins\Logging\Logger;
-use LizardsAndPumpkins\Import\RestApi\Exception\CatalogImportApiDirectoryNotReadableException;
-use LizardsAndPumpkins\Import\RestApi\Exception\CatalogImportFileNameNotFoundInRequestBodyException;
-use LizardsAndPumpkins\Import\CatalogImport;
 
-class CatalogImportApiV1PutRequestHandler extends ApiRequestHandler
+class CatalogImportApiV1PutRequestHandler extends CatalogImportApiV2PutRequestHandler
```
Comment (tim-bezhashvyly) [inline src/Import/RestApi/CatalogImportApiV1PutRequestHandler.php]: What would be a further strategy with old endpoints? Keeping them forever or announce deprecation at some point?

File: src/Import/RestApi/CatalogImportApiV1PutRequestHandler.php [hunk 3]
```diff
@@ -5,90 +5,12 @@
...
-        return new self($importDirectoryPath, $commandQueue, $logger);
-    }
-
-    final public function canProcess(HttpRequest $request) : bool
-    {
-        return HttpRequest::METHOD_PUT === $request->getMethod();
-    }
-
-    final protected function getResponse(HttpRequest $request) : HttpResponse
+    protected function createDataVersion(HttpRequest $request): DataVersion
```
Comment (tim-bezhashvyly) [inline src/Import/RestApi/CatalogImportApiV1PutRequestHandler.php]: Maybe final?

File: src/Import/RestApi/CatalogImportApiV2PutRequestHandler.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        if (!is_array($requestArguments) || !isset($requestArguments['fileName']) || !$requestArguments['fileName']) {
+            throw new CatalogImportFileNameNotFoundInRequestBodyException(
+                'Import file name is not found in request body.'
+            );
+        }
+
+        return $requestArguments['fileName'];
+    }
+
+    protected function createDataVersion(HttpRequest $request): DataVersion
```
Comment (tim-bezhashvyly) [inline src/Import/RestApi/CatalogImportApiV2PutRequestHandler.php]: Uh, this is ugly. However I have no better suggestion.

File: src/Import/RestApi/CatalogImportApiV2PutRequestHandler.php [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        $this->commandQueue = $commandQueue;
+        $this->logger = $logger;
+    }
+
+    public static function create(
+        string $importDirectoryPath,
+        CommandQueue $commandQueue,
+        Logger $logger
+    ): ApiRequestHandler {
+        if (!is_readable($importDirectoryPath)) {
```
Comment (tim-bezhashvyly) [inline src/Import/RestApi/CatalogImportApiV2PutRequestHandler.php]: Maybe also `is_dir($importDirectoryPath)`?

File: src/Import/RestApi/CatalogImportApiV2PutRequestHandler.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    private function getDataVersionFromRequest(HttpRequest $request): string
+    {
+        $requestArguments = json_decode($request->getRawBody(), true);
+
+        if (!is_array($requestArguments) ||
+            !isset($requestArguments['dataVersion']) ||
+            !$requestArguments['dataVersion']
+        ) {
```
Comment (tim-bezhashvyly) [inline src/Import/RestApi/CatalogImportApiV2PutRequestHandler.php]: This looks like a pattern (see line 87). Maybe extract it into method `$this->hasArgument($requestArguments, 'dataVersion')`?

File: src/RestApi/RestApiFactory.php [hunk 7]
```diff
@@ -33,25 +34,29 @@ private function registerApiRequestHandlers(ApiRequestHandlerLocator $requestHan
...
 
-    private function registerApiV1RequestHandlers(ApiRequestHandlerLocator $requestHandlerLocator)
+    final protected function registerApiV1RequestHandlers(ApiRequestHandlerLocator $requestHandlerLocator)
     {
-        $version = 1;
-
         $requestHandlerLocator->register(
             'put_catalog_import',
-            $version,
+            $version = 1,
```
Comment (tim-bezhashvyly) [inline src/RestApi/RestApiFactory.php]: I was thinking about using such technique more. What do you think?

File: src/RestApi/RestApiFactory.php [hunk 8]
```diff
@@ -33,25 +34,29 @@ private function registerApiRequestHandlers(ApiRequestHandlerLocator $requestHan
         $this->registerApiV1RequestHandlers($requestHandlerLocator);
     }
 
-    private function registerApiV1RequestHandlers(ApiRequestHandlerLocator $requestHandlerLocator)
+    final protected function registerApiV1RequestHandlers(ApiRequestHandlerLocator $requestHandlerLocator)
```
Comment (tim-bezhashvyly) [inline src/RestApi/RestApiFactory.php]: Hm, why?

File: tests/Integration/Suites/CatalogImportApiTest.php [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+use LizardsAndPumpkins\Http\HttpRequest;
+use LizardsAndPumpkins\Http\HttpRequestBody;
+use LizardsAndPumpkins\Http\HttpUrl;
+use LizardsAndPumpkins\Import\ImportCatalogCommand;
+use LizardsAndPumpkins\Messaging\MessageReceiver;
+use LizardsAndPumpkins\Messaging\Queue\Message;
+
+class CatalogImportApiTest extends AbstractIntegrationTest
+{
+    private function getNextMessageFromQueue(Messaging\Queue $queue): Message
```
Comment (tim-bezhashvyly) [inline tests/Integration/Suites/CatalogImportApiTest.php]: Is there a reason not to use class import?

File: tests/Unit/Suites/Import/CatalogImportWasTriggeredDomainEventTest.php [hunk 10]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+declare(strict_types = 1);
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/Import/CatalogImportWasTriggeredDomainEventTest.php]: Oh, how could I miss it.

File: src/Import/RestApi/CatalogImportApiV1PutRequestHandler.php (hunk 2 above)
Comment (Vinai) [inline src/Import/RestApi/CatalogImportApiV1PutRequestHandler.php]: Yes indeed. For now I think we just can keep it around until the data versioning has proven itself in practice. Then we can mark it as `@deprecated` with a date (+1 year maybe) and then remove it.  
The time we keep it around after deprecating it depends on the amount of maintenance cost.

File: src/Import/RestApi/CatalogImportApiV2PutRequestHandler.php (hunk 4 above)
Comment (Vinai) [inline src/Import/RestApi/CatalogImportApiV2PutRequestHandler.php]: Yes. Chose to be pragmatic until a better solution shows itself. Luckily it's only a short bit of code.

File: src/RestApi/RestApiFactory.php (hunk 8 above)
Comment (Vinai) [inline src/RestApi/RestApiFactory.php]: To make it testable. I wanted to be able to test what handlers get registered before I register a new one. That wasn't covered so far, and I think it's rather important.

File: src/RestApi/RestApiFactory.php (hunk 7 above)
Comment (Vinai) [inline src/RestApi/RestApiFactory.php]: Whatever is clearer. It depends on each case.

File: tests/Integration/Suites/CatalogImportApiTest.php (hunk 9 above)
Comment (Vinai) [inline tests/Integration/Suites/CatalogImportApiTest.php]: Ah no, it's just what PHPStorm does.

**Comment by Vinai** [comment]
Ready for next review.

**Pull request merged**